lapis/storage.py

```diff
@@ -0,0 +1,223 @@
+        stored_file.increment_accesses()
+        print(
+            "UPDATE: Job {}, File {} @ {}".format(
+                job_repr, stored_file.filename, time.now
+            )
+        )
+
+    async def transfer(self, file, job_repr):
+        """
+        Manages file transfer via the storage elements connection and updates file
+        information. If the file should have been deleted since it was originally
+        looked up the resulting error is not raised.
+        :param file:
+        :param job_repr:  Needed for debug output, will be replaced
+        :return:
+        """
+        await self.connection.transfer(file.filesize)
+        try:
+            await self.update_file(self.find_file(file.filename), job_repr)
+        except AttributeError:
+            pass
+
+    async def apply_caching_decision(self, requested_file: RequestedFile, job_repr):
+        """
+        Applies the storage objects caching algorithm to the requested_file and
+        initiates resulting changes like placement and deletion of files
+        :param requested_file:
+        :param job_repr: Needed for debug output, will be replaced
+        :return:
+        """
+
+        print(
+            "APPLY CACHING DECISION: Job {}, File {} @ {}".format(
+                job_repr, requested_file.filename, time.now
+            )
+        )
+        to_be_removed = self.cachealgorithm.consider(requested_file)
+        if not to_be_removed:
+            await self.add_to_storage(requested_file, job_repr)
+        elif to_be_removed == {requested_file}:
+            # file will not be cached because it either does not match
+            # conditions or because there is no space in the cache
+            print(
+                "APPLY CACHING DECISION: Job {}, File {}: File wasnt "
+                "cached @ {}".format(job_repr, requested_file.filename, time.now)
+            )
+        else:
+            for file in to_be_removed:
+                await self.remove_from_storage(file, job_repr)
+
+    async def look_up_file(self, requested_file: RequestedFile, queue: Queue, job_repr):
+        """
+        Searches storage object for the requested_file and sends result (amount of
+        cached data, storage object) to queue if queue was passed as parameter.
+        If no queue was passed the result is returned normally. This might be needed
+        when looking up files during coordination and is to be removed if it's not
+        necessary.
+        :param requested_file:
+        :param queue:
+        :param job_repr: Needed for debug output, will be replaced
+        :return: (amount of cached data, storage object)
+        """
+        print(
+            "LOOK UP FILE: Job {}, File {}, Storage {} @ {}".format(
+                job_repr, requested_file.filename, repr(self), time.now
+            )
+        )
+
+        class LookUpInformation(NamedTuple):
+            cached_filesize: int
+            storage: Storage
+
+        if queue:
```

Comment:
I strongly recommend to separate concurrent (queue) and non-concurrent (``return``) functionality. Do not decide based on the value of inputs whether it is one or the other. Otherwise, people (including future you) have to know how this function works internally and all its arguments in order to use it properly.

In this specific case, it appears you can drop the concurrent/queue and just directly return the result for now.